File: typescript/src/application/usecase/order/additionalBuyOrderUsecase.ts
import Decimal from "decimal.js";
import { AccountRepository } from "../../repository/accountRepository";
import { PortfolioRepository } from "../../repository/portfolioRepository";
import { AppConstants } from "../../../domain/appConstants";
import { UserId } from "../../../domain/userId";

export interface AdditionalBuyOrderUsecaseInput {
  userId: UserId;
  amount: Decimal;
}

export class AdditionalBuyOrderUsecaseException extends Error {}
export class AdditionalBuyUserNotFoundException extends AdditionalBuyOrderUsecaseException {
  constructor() {
    super("user not found");
  }
}
export class AdditionalBuyAmountTooSmallException extends AdditionalBuyOrderUsecaseException {
  constructor() {
    super("amount too small");
  }
}

export class AdditionalBuyOrderUsecase {
  constructor(
    private readonly accountRepository: AccountRepository,
    private readonly portfolioRepository: PortfolioRepository,
  ) {}

  async run(input: AdditionalBuyOrderUsecaseInput): Promise<void> {
    if (input.amount.lessThan(AppConstants.minOperationAmount)) {
      throw new AdditionalBuyAmountTooSmallException();
    }
    const account = await this.accountRepository.find(input.userId);
    if (account === undefined) {
      throw new AdditionalBuyUserNotFoundException();
    }
    const portfolio = await this.portfolioRepository.get();
    const updated = account.addFunds(input.amount, portfolio);
    await this.accountRepository.upsert(input.userId, updated);
  }
}

File: golang/internal/application/usecase/order/additional_buy_order_usecase.go
package order

import (
	"errors"

	"folio/codinginterview/internal/application/repository"
	"folio/codinginterview/internal/domain"

	"github.com/shopspring/decimal"
)

var (
	ErrAdditionalBuyUserNotFound   = errors.New("user has no live account")
	ErrAdditionalBuyAmountTooSmall = errors.New("amount is too small")
)

type AdditionalBuyOrderUsecaseInput struct {
	UserId domain.UserId
	Amount decimal.Decimal
}

type AdditionalBuyOrderUsecase struct {
	accountRepo   repository.AccountRepository
	portfolioRepo repository.PortfolioRepository
}

func NewAdditionalBuyOrderUsecase(
	accountRepo repository.AccountRepository,
	portfolioRepo repository.PortfolioRepository,
) *AdditionalBuyOrderUsecase {
	return &AdditionalBuyOrderUsecase{
		accountRepo:   accountRepo,
		portfolioRepo: portfolioRepo,
	}
}

func (u *AdditionalBuyOrderUsecase) Run(input AdditionalBuyOrderUsecaseInput) error {
	if input.Amount.LessThan(domain.MinOperationAmount) {
		return ErrAdditionalBuyAmountTooSmall
	}

	account, err := u.accountRepo.Find(input.UserId)
	if err != nil {
		return err
	}
	if account == nil {
		return ErrAdditionalBuyUserNotFound
	}

	portfolio, err := u.portfolioRepo.Get()
	if err != nil {
		return err
	}

	updated := account.AddFunds(input.Amount, portfolio)

	return u.accountRepo.Upsert(input.UserId, updated)
}
